text/0000-unnamed-fields.md

```diff
@@ -0,0 +1,350 @@
+
+This pattern appears throughout many C APIs. The Windows and POSIX APIs both
+use this pattern extensively. However, Rust currently can't represent this
+pattern in a straightforward way. While Rust supports structs and unions, every
+such struct and union must have a field name. When creating a binding to such
+an interface, whether manually or using a binding generator, the binding must
+invent an artificial field name that does not appear in the original interface.
+
+This RFC proposes a minimal mechanism to support such interfaces in Rust.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This explanation should appear after the definition of `union`, and after an
+explanation of the rationale for `union` versus `enum` in Rust.
+
+Please note that most Rust code will want to use an `enum` to define types that
+contain a discriminant and various disjoint fields. The unnamed field mechanism
+here exist primarily for compatibility with interfaces defined by non-Rust
+languages, such as C. Types declared with this mechanism require `unsafe` code
+to access.
+
+A `struct` defines a set of fields all available at the same time, with storage
+available for each. A `union` defines (in an unsafe, unchecked manner) a set of
+mutually exclusive fields, with overlapping storage. Some types and interfaces
+may require nesting such groupings. For instance, a `struct` may contain a set
+of common fields and a `union` of fields needed for different variations of the
+structure; conversely, a `union` contain a `struct` grouping together fields
+needed simultaneously.
+
+Such groupings, however, do not always have associated types and names. A
+structure may contain groupings of fields where the fields have meaningful
+names, but the groupings of fields do not. In this case, the structure can
+contain *unnamed fields* of `struct` or `union` type, to group the fields
+together, and determine which fields overlap.
+
+As an example, when defining a `struct`, you may have a set of fields that will
+never be used at the same time, so you could overlap the storage of those
+fields. This pattern often occurs within C APIs, when defining an interface
+similar to a Rust `enum`. You could do so by declaring a separate `union` type
+and a field of that type. With the unnamed fields mechanism, you can also
+define an unnamed grouping of overlapping fields inline within the `struct`,
+using the `union` keyword:
+
+```rust
+struct S {
+    a: u32,
+    union {
+        b: u32,
+        c: f32,
+    },
+    d: u64,
+}
+```
+
+Given a struct `s` of this type, code can access `s.a`, `s.d`, and either `s.b`
+or `s.c`. Accesses to `a` and `d` can occur in safe code; accesses to `b` and
+`c` require unsafe code, and `b` and `c` overlap, requiring care to access only
+the field whose contents make sense at the time. As with any `union`, code
+cannot borrow `s.b` and `s.c` simultaneously.
+
```

Comment:
> Given a struct `s` of this type, code can access `s.a`, `s.d`, and either `s.b`
> or `s.c`. Accesses to `a` and `d` can occur in safe code; accesses to `b` and
> `c` require unsafe code

Because of the `siginfo_t` example, where different implementations of POSIX put different subsets of the standardized fields into an anonymous union, I recommend that this be changed: if a struct has an anonymous union inside, it should become unsafe to access _any_ of its fields.